### PR TITLE
feat!: Refactor module to align with current standards, add missing addtional Redshift resources to module

### LINKED ISSUE #67: Allow settings the default role
## Is your request related to a new offering from AWS?

4.16.0 added argument default_iam_role_arn to aws_redshift_cluster

## Is your request related to a problem? Please describe.

Creating my redshifdt cluster, I need to associate a default role allowing the COPY (from S3 bucket to redshift table) command. I can associate the said role but not set it as default. 

## Describe the solution you'd like.

Add `default_iam_role_arn` var

## Describe alternatives you've considered.

Clicking in the UI for now

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- Minimum supported version of Terraform AWS provider updated to v4.17 to support latest resources
- Minimum supported version of Terraform updated to v1.0
- `create` variable added to control whether all resources are created or not. This means that all resources now use the zeroth index `[0]` in the resource name
- Support for generating a random password for the `master_password`
- `aws_redshift_snapshot_schedule` resource including the variables and outputs used to support it
- `aws_redshift_snapshot_schedule_association` resource including the variables and outputs used to support it
- `aws_redshift_scheduled_action` resource including support for creating the IAM role and policies plus the associated variables and outputs to support
- `aws_redshift_usage_limit` resource including the variables and outputs used to support it
- `aws_redshift_authentication_profile` resource including the variables and outputs used to support it
- `aws_redshift_cluster_iam_roles` resource including the variables and outputs used to support it

See https://github.com/bryantbiggs/terraform-aws-redshift/blob/chore/updates/UPGRADE-4.0.md for comprehensive list of changes

## Motivation and Context
- Aligns module with current standards and practices
- Updates Terraform and AWS provider versions to latest GA versions that support features used
- Adds resources and functionality that have since been added to the provider
- Closes #66 
- Closes #67 

## Breaking Changes
- Yes! The module has been refactored to utilize count for conditional creation which renders this change as breaking. See https://github.com/bryantbiggs/terraform-aws-redshift/blob/chore/updates/UPGRADE-4.0.md for comprehensive list of changes

## How Has This Been Tested?
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

### CODE REVIEW

**Comment by bryantbiggs** [comment]
Shhhh

**Comment by DrFaust92** [comment]
LGTM

**Comment by bryantbiggs** [comment]
Checks will fail until (hopefully) tomorrow when v4.16 is released

**Comment by antonbabenko** [comment]
I think #66 is somewhat about the same as this PR - refactoring and adding more resources.

@bryantbiggs could you please verify/extend it so we can merge one and close another?

**Comment by bryantbiggs** [comment]
Yes, it looks like most of what I had here was copied over there. I'll finish up this PR today and we should be able to merge this week after the next provider release

**Comment by Zvi-Prycto** [comment]
@bryantbiggs Yeah, I used your PR as a base, the master_password was failing for me, so I removed it and added the missing things I needed :D (mostly the AZ relocation).

**Comment by DrFaust92** [comment]
FYI I hope a few more resources and changes and cover most of redshifts API in the coming days

**Comment by antonbabenko** [comment]
@DrFaust92 You always know something we don't know until it is released! Thank you for that! ;)

**Comment by bryantbiggs** [comment]
I'm a bit perplexed - I deploy the `Before` configuration, then swap out the `After` configuration and I don't have to make any `terraform state mv ...` commands despite the fact that `count` is now used and the state resource entry is now changed with the 0 index addition. Did I miss something?!

### Common/Shared

```hcl
provider "aws" {
  region = "eu-west-1"
}

locals {
  name = "my-cluster"
}

module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.0"

  name             = local.name
  cidr             = "10.10.0.0/16"
  azs              = ["eu-west-1a", "eu-west-1b", "eu-west-1c"]
  redshift_subnets = ["10.10.41.0/24", "10.10.42.0/24", "10.10.43.0/24"]
}

module "sg" {
  source  = "terraform-aws-modules/security-group/aws//modules/redshift"
  version = "~> 4.0"

  name   = "demo-redshift"
  vpc_id = module.vpc.vpc_id

  ingress_cidr_blocks = [module.vpc.vpc_cidr_block]
  egress_rules        = ["all-all"]
}
```

### Before

```hcl
module "redshift" {
  source  = "terraform-aws-modules/redshift/aws"
  version = "3.4.1"

  cluster_identifier      = local.name
  cluster_node_type       = "dc2.large"
  cluster_number_of_nodes = 1

  cluster_database_name   = "mydb"
  cluster_master_username = "mydbuser"
  cluster_master_password = "MySecretPassw0rd"

  subnets                = module.vpc.redshift_subnets
  vpc_security_group_ids = [module.sg.security_group_id]
}
```

### After

```hcl
module "redshift" {
  source = "../../"

  cluster_identifier = local.name
  node_type          = "dc2.large"
  number_of_nodes    = 1

  database_name   = "mydb"
  master_username = "mydbuser"
  master_password = "MySecretPassw0rd"

  subnet_ids             = module.vpc.redshift_subnets
  vpc_security_group_ids = [module.sg.security_group_id]

  # Maintain v3.x settings
  encrypted                           = false
  automated_snapshot_retention_period = 0
  parameter_group_name                = "${local.name}-redshift-1-0-custom-params"
  parameter_group_parameters = {
    wlm_json_configuration = {
      name = "wlm_json_configuration"
      value = jsonencode([
        {
          query_concurrency = 5
        }
      ])
    }
    require_ssl = {
      name  = "require_ssl"
      value = false
    }
    use_fips_ssl = {
      name  = "use_fips_ssl"
      value = false
    }
    enable_user_activity_logging = {
      name  = "enable_user_activity_logging"
      value = false
    }
    max_concurrency_scaling_clusters = {
      name  = "max_concurrency_scaling_clusters"
      value = 1
    }
    enable_case_sensitive_identifier = {
      name  = "enable_case_sensitive_identifier"
      value = false
    }
  }
  subnet_group_description = "Redshift subnet group of ${local.name}"
  create_random_password   = false
}
```

File: UPGRADE-4.0.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+- `aws_redshift_scheduled_action` resource including support for creating the IAM role and policies plus the associated variables and outputs to support
+- `aws_redshift_usage_limit` resource including the variables and outputs used to support it
+- `aws_redshift_authentication_profile` resource including the variables and outputs used to support it
+- `aws_redshift_cluster_iam_roles` resource including the variables and outputs used to support it
+
+### Modified
+
+- `number_of_nodes` default value of `3` changed to `1`
+- `cluster_version` default value of `"1.0"` changed to `null`
+- `cluster_node_type` default value of `dc2.large` added
```
Comment (antonbabenko) [inline UPGRADE-4.0.md:27]: Should `cluster_node_type` be up to a user to specify and avoid default values?

File: examples/complete/main.tf [hunk 2]
```diff
@@ -1,55 +1,343 @@
...
+
+################################################################################
+# Default
+################################################################################
+
+module "default" {
+  source = "../../"
+
+  cluster_identifier = "${local.name}-default"
+
```
Comment (antonbabenko) [inline examples/complete/main.tf:206]: Let's make it clear for a user with specified `node_type`

File: examples/complete/main.tf [hunk 3]
```diff
@@ -1,55 +1,343 @@
...
+resource "random_pet" "s3_bucket" {
+  length = 2
+}
+
+module "s3_logs" {
+  source  = "terraform-aws-modules/s3-bucket/aws"
+  version = "~> 3.0"
+
+  bucket = "${local.name}-${random_pet.s3_bucket.id}"
```
Comment (antonbabenko) [inline examples/complete/main.tf:321]:
```suggestion
  bucket = "${local.name}-${random_pet.this.id}"
```

**Review by antonbabenko — COMMENTED** [review]
Looks pretty good. Just some minor comments here and there.

File: examples/complete/main.tf [hunk 4]
```diff
@@ -1,55 +1,343 @@
...
+################################################################################
+# Disabled
+################################################################################
+
+module "disabled" {
+  source = "../../"
+
+  create = false
+
+  cluster_identifier = local.name
```
[the remark this answers is not in the document]
Comment (bryantbiggs) [reply examples/complete/main.tf]: yes - removed

File: examples/complete/main.tf (hunk 2 above)
Comment (bryantbiggs) [reply examples/complete/main.tf:208]: done

File: UPGRADE-4.0.md (hunk 1 above)
Comment (bryantbiggs) [reply UPGRADE-4.0.md]: yes - not sure what I was thinking there. removed

**Comment by antonbabenko** [comment]
> I'm a bit perplexed - I deploy the `Before` configuration, then swap out the `After` configuration and I don't have to make any `terraform state mv ...` commands despite the fact that `count` is now used and the state resource entry is now changed with the 0 index addition. Did I miss something?!

I think you are right. The changes should require `terraform state mv ...` command. I didn't run examples for real, just in my head.

**Comment by bryantbiggs** [comment]
> > I'm a bit perplexed - I deploy the `Before` configuration, then swap out the `After` configuration and I don't have to make any `terraform state mv ...` commands despite the fact that `count` is now used and the state resource entry is now changed with the 0 index addition. Did I miss something?!
> 
> I think you are right. The changes should require `terraform state mv ...` command. I didn't run examples for real, just in my head.

I ran them a couple times (what I posted above) a couple times and on two different machines - same result. We can always update the upgrade guide if users report different results

**Comment by antonbabenko** [comment]
This PR is included in version 4.0.0 :tada:

**Pull request merged**